src/hal/arm/string.h

```diff
@@ -117,10 +107,10 @@ static inline unsigned int hal_strlen(const char *s)
 	(" \
 	1: \
 		ldrb r1, [%1, %0]; \
-		add %0, #1; \
-		cmp r1, #0; \
-		bne 1b; \
-		sub %0, #1;"
+		cbz r1, 2f; \
+		adds %0, #1; \
```

Comment:
This changes code behaviour. New code outputs one less.

Reply:
The value returned by the function is valid (the code is almost identical to the one generated by the compiler).
disassembly strlen in c:
```
00000000 <strlen>:
   0:   4603            mov     r3, r0
   2:   2000            movs    r0, #0
   4:   5c1a            ldrb    r2, [r3, r0]
   6:   b902            cbnz    r2, a <strlen+0xa>
   8:   4770            bx      lr
   a:   3001            adds    r0, #1
   c:   e7fa            b.n     4 <strlen+0x4>
```

Reply:
I tested it, result is the same indeed. I missed sub on the previous version.
I think that there is no need for _adds_ instruction, _add_ if enough. We don't need add result's flags, are we?